src/herder/TxQueueLimiter.h

```diff
@@ -13,51 +13,14 @@
 namespace stellar
 {
 
-class SingleTxStack : public TxStack
-{
-  public:
-    SingleTxStack(TransactionFrameBasePtr tx) : mTx(tx)
-    {
-    }
-
-    TransactionFrameBasePtr
-    getTopTx() const override
-    {
-        releaseAssert(mTx);
-        return mTx;
-    }
-
-    void
-    popTopTx() override
-    {
-        releaseAssert(mTx);
-        mTx = nullptr;
-    }
-
-    bool
-    empty() const override
-    {
-        return mTx == nullptr;
-    }
-
-    Resource
-    getResources() const override
-    {
-        releaseAssert(mTx);
-        return Resource(mTx->getResources(/* useByteLimitInClassic */ false));
-    }
-
-  private:
-    TransactionFrameBasePtr mTx;
-};
-
 class TxQueueLimiter
 {
     // number of ledgers we can pool in memory
     uint32 const mPoolLedgerMultiplier;
     LedgerManager& mLedgerManager;
 
-    UnorderedMap<TransactionFrameBasePtr, TxStackPtr> mStackForTx;
+    UnorderedMap<TransactionFrameBasePtr, SurgePricingQueueTxBasePtr>
```

Comment:
isn't this whole data structure redundant now?

Reply:
It's still necessary because we need to remove containers from the queue using the tx pointer. Inheritance would've helped here, but I still think it's the lesser evil.

Reply:
There are several users of custom TxContainer implementations (2 now, 1 more for necessary parallel tx building), but these can be reasonably simply replaced with the maps. But in the end this seems to reduce the line count, so the change is probably for the better.
FWIW surge pricing queue is not quite an iterator, it's also a container (as in, it has to store data), so the fundamental issue is whether it allows storing data in a more or less flexible fashion. I think using more rigid interface + maps is fine.